cpp/src/arrow/engine/substrait/relation_internal.cc

```diff
@@ -88,35 +142,41 @@ Result<compute::Declaration> FromProto(const substrait::Rel& rel,
 
       std::shared_ptr<dataset::FileFormat> format;
       auto filesystem = std::make_shared<fs::LocalFileSystem>();
-      std::vector<std::shared_ptr<dataset::FileFragment>> fragments;
+      std::vector<fs::FileInfo> files;
 
       for (const auto& item : read.local_files().items()) {
-        if (item.path_type_case() !=
-            substrait::ReadRel_LocalFiles_FileOrFiles::kUriFile) {
-          return Status::NotImplemented(
-              "substrait::ReadRel::LocalFiles::FileOrFiles with "
-              "path_type other than uri_file");
+        std::string path;
+        if (item.path_type_case() ==
+            substrait::ReadRel_LocalFiles_FileOrFiles::kUriPath) {
+          path = item.uri_path();
+        } else if (item.path_type_case() ==
+                   substrait::ReadRel_LocalFiles_FileOrFiles::kUriFile) {
+          path = item.uri_file();
+        } else if (item.path_type_case() ==
+                   substrait::ReadRel_LocalFiles_FileOrFiles::kUriFolder) {
+          path = item.uri_folder();
+        } else {
+          path = item.uri_path_glob();
         }
 
         if (item.format() ==
             substrait::ReadRel::LocalFiles::FileOrFiles::FILE_FORMAT_PARQUET) {
           format = std::make_shared<dataset::ParquetFileFormat>();
-        } else if (util::string_view{item.uri_file()}.ends_with(".arrow")) {
+        } else if (util::string_view{path}.ends_with(".arrow")) {
```

Comment:
This seems a bit incorrect (e.g. the format is parquet if my glob is `dataset/files/*.arrow` but not `dataset/files/*`) but this particular branch is a bit of a hack until https://github.com/substrait-io/substrait/pull/169 merges anyways.

Reply:
So just to be clear, I'm not suggesting a change here.